client/file-browser/browse-files.jsx

```diff
@@ -0,0 +1,526 @@
+      // Remove the last entry if it's empty (due to a trailing slash)
+      pieces.pop()
+    }
+    const { elems } = pieces.reduce(
+      (r, piece, i) => {
+        const isLast = i === pieces.length - 1
+        r.curPath += (i === 0 ? '' : '\\') + piece
+        // Save the value at the current time so the function doesn't always use the last value
+        const navPath = r.curPath
+        r.elems.push(
+          <span
+            key={i}
+            className={isLast ? styles.breadcrumbActive : styles.breadcrumb}
+            onClick={isLast ? undefined : () => this.props.onNavigate(navPath)}>
+            {piece}
+          </span>,
+        )
+        r.elems.push(<ChevronRight key={i + '|'} className={styles.breadcrumbSeparator} />)
+
+        return r
+      },
+      { elems: [], curPath: '' },
+    )
+
+    return <div className={this.props.className}>{elems}</div>
+  }
+}
+
+@connect(state => ({ fileBrowser: state.fileBrowser }))
+export default class Files extends React.Component {
+  state = {
+    focusedPath: window.localStorage.getItem(this.props.browseId + FOCUSED_KEY),
+  }
+
+  focusTimeout = null
+  browserRef = React.createRef()
+  contentRef = React.createRef()
+  listRef = React.createRef()
+
+  getEntries = memoize(
+    props => {
+      const { browseId, fileBrowser, root, fileTypes } = props
+      const { path, files, folders } = fileBrowser[browseId]
+      const isRootFolder = path === root
+      const upOneDir = isRootFolder ? new List([]) : new List([{ type: 'up', path: path + '\\..' }])
+      const filteredFiles = files.filter(f => fileTypes[f.extension])
+      return upOneDir.concat(folders, filteredFiles)
+    },
+    (newArgs, prevArgs) => {
+      const { browseId, fileBrowser } = newArgs[0]
+      const { path, folders, files } = fileBrowser[browseId]
+      const { browseId: prevBrowseId, fileBrowser: prevFileBrowser } = prevArgs[0]
+      const { path: prevPath, folders: prevFolders, files: prevFiles } = prevFileBrowser[
+        prevBrowseId
+      ]
+
+      return path === prevPath && folders.size === prevFolders.size && files.size === prevFiles.size
+    },
+  )
+
+  _saveToLocalStorage = () => {
+    const { focusedPath } = this.state
+
+    if (focusedPath) {
+      window.localStorage.setItem(this.props.browseId + FOCUSED_KEY, focusedPath)
+    }
+  }
+
+  componentDidMount() {
+    const { browseId, root } = this.props
+    const { focusedPath } = this.state
+    const { path } = this.props.fileBrowser[browseId]
+    if (path === '') {
+      const initialPath = focusedPath ? pathApi.parse(focusedPath).dir : root
+
+      this.props.dispatch(changePath(browseId, initialPath))
+    } else {
+      this.props.dispatch(getFiles(browseId, path))
+    }
+
+    // Focus *something* when browser is opened, because if we don't, whatever was focused before
+    // the browser was opened will still have focus and will mess with our keyboard events.
+    this.focusTimeout = setTimeout(() => {
```

Comment:
I'd probably prefer `Promise.resolve().then(() => { ... })` over a `setTimeout(0)`

(I'm guessing you probably took this from something I wrote, soooo :P)

IIRC we may have had some kind of actual timeout before focusing anything because a focus animation caused jank in panel opening? But maybe that doesn't matter here. If it doesn't matter, you could probably just focus an element when the ref is set

Reply:
Yeah, using a callback ref instead of `createRef` API might be better here, as we can focus the element as soon the React calls the ref setter.